File: mapper/src/main/java/com/azure/autorest/template/ModelTemplate.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

package com.azure.autorest.template;

import com.azure.autorest.extension.base.plugin.JavaSettings;
import com.azure.autorest.model.clientmodel.ArrayType;
import com.azure.autorest.model.clientmodel.ClientModel;
import com.azure.autorest.model.clientmodel.ClientModelProperty;
import com.azure.autorest.model.clientmodel.IType;
import com.azure.autorest.model.clientmodel.ListType;
import com.azure.autorest.model.javamodel.JavaFile;
import com.azure.autorest.model.javamodel.JavaModifier;

import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

/** 
 Writes a ClientModel to a JavaFile.
*/
public class ModelTemplate implements IJavaTemplate<ClientModel, JavaFile>
{
    private static ModelTemplate _instance = new ModelTemplate();
    public static ModelTemplate getInstance()
    {
        return _instance;
    }

    private ModelTemplate()
    {
    }

    public final void write(ClientModel model, JavaFile javaFile)
    {
        JavaSettings settings = JavaSettings.getInstance();
        Set<String> imports = new HashSet<String>();
        model.addImportsTo(imports, settings);

        javaFile.declareImport(imports);

        javaFile.javadocComment(settings.getMaximumJavadocCommentWidth(), (comment) ->
        {
                comment.description(model.getDescription());
        });

        boolean hasDerivedModels = !model.getDerivedModels().isEmpty();
        if (model.getIsPolymorphic())
        {
            javaFile.annotation(String.format("JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = \"%1$s\"%2$s)", model.getPolymorphicDiscriminator(), (hasDerivedModels ? String.format(", defaultImpl = %1$s.class", model.getName()) : "")));
            javaFile.annotation(String.format("JsonTypeName(\"%1$s\")", model.getSerializedName()));

            if (hasDerivedModels)
            {
                javaFile.line("@JsonSubTypes({");
                javaFile.indent(() -> {
                        Function<ClientModel, String> getDerivedTypeAnnotation = (ClientModel derivedType) -> String.format("@JsonSubTypes.Type(name = \"%1$s\", value = %2$s.class)", derivedType.getSerializedName(), derivedType.getName());

                        for (int i = 0; i != model.getDerivedModels().size() - 1; i++)
                        {
                            ClientModel derivedModel = model.getDerivedModels().get(i);
                            javaFile.line(getDerivedTypeAnnotation.apply(derivedModel) + ',');
                        }
                        javaFile.line(getDerivedTypeAnnotation.apply(model.getDerivedModels().get(model.getDerivedModels().size() - 1)));
                });
                javaFile.line("})");
            }
        }

        if (settings.getShouldGenerateXmlSerialization())
        {
            javaFile.annotation(String.format("JacksonXmlRootElement(localName = \"%1$s\")", model.getXmlName()));
        }

        if (model.getNeedsFlatten())
        {
            javaFile.annotation("JsonFlatten");
        }

        ArrayList<JavaModifier> classModifiers = new ArrayList<JavaModifier>();
        if (!hasDerivedModels && !model.getNeedsFlatten())
        {
            classModifiers.add(JavaModifier.Final);
        }

        String classNameWithBaseType = model.getName();
        if (model.getParentModel() != null)
        {
            classNameWithBaseType += String.format(" extends %1$s", model.getParentModel().getName());
        }
        javaFile.annotation("Fluent");
        javaFile.publicClass(classModifiers, classNameWithBaseType, (classBlock) ->
            {
                Function<ClientModelProperty, String> propertyXmlWrapperClassName = (ClientModelProperty property) -> property.getXmlName() + "Wrapper";

                for (ClientModelProperty property : model.getProperties())
                {
                    String xmlWrapperClassName = propertyXmlWrapperClassName.apply(property);
                    if (settings.getShouldGenerateXmlSerialization() && property.getIsXmlWrapper())
                    {
                        classBlock.privateStaticFinalClass(xmlWrapperClassName, innerClass ->
                        {
                            IType propertyClientType = property.getWireType().getClientType();

                            innerClass.annotation(String.format("JacksonXmlProperty(localName = \"%1$s\")", property.getXmlListElementName()));
                            innerClass.privateFinalMemberVariable(propertyClientType.toString(), "items");

                            innerClass.annotation("JsonCreator");
                            innerClass.privateConstructor(String.format("%1$s(@JacksonXmlProperty(localName = \"%2$s\") %3$s items)", xmlWrapperClassName, property.getXmlListElementName(), propertyClientType), constructor -> constructor.line("this.items = items;"));
                        });
                    }

                    classBlock.blockComment(settings.getMaximumJavadocCommentWidth(), (comment) ->
                    {
                        comment.line(property.getDescription());
                    });

                    if (property.getHeaderCollectionPrefix() != null && !property.getHeaderCollectionPrefix().isEmpty())
                    {
                        classBlock.annotation("HeaderCollection(\"" + property.getHeaderCollectionPrefix() + "\")");
                    }
                    else if (settings.getShouldGenerateXmlSerialization() && property.getIsXmlAttribute())
                    {
                        String localName = settings.getShouldGenerateXmlSerialization() ? property.getXmlName() : property.getSerializedName();
                        classBlock.annotation(String.format("JacksonXmlProperty(localName = \"%1$s\", isAttribute = true)", localName));
                    }
                    else if (settings.getShouldGenerateXmlSerialization() && property.getWireType() instanceof ListType && !property.getIsXmlWrapper())
                    {
                        classBlock.annotation(String.format("JsonProperty(\"%1$s\")", property.getXmlListElementName()));
                    }
                    else if (property.getAnnotationArguments() != null && !property.getAnnotationArguments().isEmpty())
                    {
                        classBlock.annotation(String.format("JsonProperty(%1$s)", property.getAnnotationArguments()));
                    }

                    if (settings.getShouldGenerateXmlSerialization())
                    {
                        if (property.getIsXmlWrapper())
                        {
                            classBlock.privateMemberVariable(String.format("%1$s %2$s", xmlWrapperClassName, property.getName()));
                        }
                        else if (property.getWireType() instanceof ListType)
                        {
                            classBlock.privateMemberVariable(String.format("%1$s %2$s = new ArrayList<>()", property.getWireType(), property.getName()));
                        }
                        else
                        {
                            classBlock.privateMemberVariable(String.format("%1$s %2$s", property.getWireType(), property.getName()));
                        }
                    }
                    else
                    {
                        classBlock.privateMemberVariable(String.format("%1$s %2$s", property.getWireType(), property.getName()));
                    }
                }

                List<ClientModelProperty> constantProperties = model.getProperties().stream().filter(ClientModelProperty::getIsConstant).collect(Collectors.toList());
                if (!constantProperties.isEmpty())
                {
                    classBlock.javadocComment(settings.getMaximumJavadocCommentWidth(), (comment) ->
                    {
                        comment.description(String.format("Creates an instance of %1$s class.", model.getName()));
                    });
                    classBlock.publicConstructor(String.format("%1$s()", model.getName()), (constructor) ->
                    {
                        for (ClientModelProperty constantProperty : constantProperties)
                        {
                            constructor.line(String.format("%1$s = %2$s;", constantProperty.getName(), constantProperty.getDefaultValue()));
                        }
                    });
                }

                for (ClientModelProperty property : model.getProperties())
                {
                    IType propertyType = property.getWireType();
                    IType propertyClientType = propertyType.getClientType();

                    classBlock.javadocComment(settings.getMaximumJavadocCommentWidth(), (comment) ->
                    {
                        comment.description(String.format("Get the %1$s property: %2$s", property.getName(), property.getDescription()));
                        comment.methodReturns(String.format("the %1$s value", property.getName()));
                    });
                    classBlock.publicMethod(String.format("%1$s %2$s()", propertyClientType, property.getName()), (methodBlock) ->
                    {
                        String sourceTypeName = propertyType.toString();
                        String targetTypeName = propertyClientType.toString();
                        String expression = "this.{property.Name}";
                        if (propertyClientType.equals(ArrayType.ByteArray))
                        {
                            expression = "ImplUtils.clone({expression})";
                        }
                        if (sourceTypeName.equals(targetTypeName))
                        {
                            if (settings.getShouldGenerateXmlSerialization() && property.getIsXmlWrapper() && property.getWireType() instanceof ListType)
                            {
                                methodBlock.ifBlock("this.{property.Name} == null", ifBlock ->
                                        ifBlock.line("this.{property.Name} = new {propertyXmlWrapperClassName(property)}(new ArrayList<{listType.ElementType}>());"));
                                methodBlock.methodReturn("this.{property.Name}.items");
                            }
                            else
                            {
                                methodBlock.methodReturn("{expression}");
                            }
                        }
                        else
                        {
                            methodBlock.ifBlock("{expression} == null", (ifBlock) -> ifBlock.methodReturn("null"));

                            String propertyConversion = propertyType.convertToClientType(expression);

                            methodBlock.methodReturn(propertyConversion);
                        }
                    });

                    if (!property.getIsReadOnly())
                    {
                        classBlock.javadocComment(settings.getMaximumJavadocCommentWidth(), (comment) ->
                                {
                                        comment.description("Set the {property.Name} property: {property.Description}");
                        comment.param(property.getName(), "the {property.Name} value to set");
                        comment.methodReturns("the {model.Name} object itself.");
                        });
                        classBlock.publicMethod("{model.Name} {property.Name.ToCamelCase()}({propertyClientType} {property.Name})", (methodBlock) -> {
                                String expression;
                                if (propertyClientType.equals(ArrayType.ByteArray)) {
                                    expression = "ImplUtils.clone({expression})";
                                } else {
                                    expression = property.getName();
                                }
                                if (propertyClientType != propertyType) {
                                    methodBlock.ifBlock("{property.Name} == null", (ifBlock) -> ifBlock.line("this.{property.Name} = null;"))
                                        .elseBlock((elseBlock) -> {
                                                String sourceTypeName = propertyClientType.toString();
                                                String targetTypeName = propertyType.toString();
                                                String propertyConversion = propertyType.convertFromClientType(expression);
                                                elseBlock.line("this.{property.Name} = {propertyConversion};");
                                            });
                                } else {
                                    if (settings.getShouldGenerateXmlSerialization() && property.getIsXmlWrapper()) {
                                        methodBlock.line("this.{property.Name} = new {propertyXmlWrapperClassName(property)}({expression});");
                                    } else {
                                        methodBlock.line("this.{property.Name} = {expression};");
                                    }
                                }
                                methodBlock.methodReturn("this");
                            });
                    }
                }
            });
    }
}

File: src/Template/ModelTemplate.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Net;
using System.Text;
using AutoRest.Core;
using AutoRest.Core.Utilities;
using AutoRest.Extensions;
using AutoRest.Core.Model;
using Newtonsoft.Json;
using AutoRest.Core.Utilities.Collections;
using Newtonsoft.Json.Linq;
using AutoRest.Java.Model;
using System.Text.RegularExpressions;

namespace AutoRest.Java
{
    /// <summary>
    /// Writes a ClientModel to a JavaFile.
    /// </summary>
    public class ModelTemplate : IJavaTemplate<ClientModel, JavaFile>
    {
        private static ModelTemplate _instance = new ModelTemplate();
        public static ModelTemplate Instance => _instance;

        private ModelTemplate()
        {
        }

        public void Write(ClientModel model, JavaFile javaFile)
        {
            var settings = JavaSettings.Instance;
            ISet<string> imports = new HashSet<string>();
            model.AddImportsTo(imports, settings);

            javaFile.Import(imports);

            javaFile.JavadocComment(settings.MaximumJavadocCommentWidth, (comment) =>
            {
                comment.Description(model.Description);
            });

            bool hasDerivedModels = model.DerivedModels.Any();
            if (model.IsPolymorphic)
            {
                javaFile.Annotation($"JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = \"{model.PolymorphicDiscriminator}\"{(hasDerivedModels ? $", defaultImpl = {model.Name}.class" : "")})");
                javaFile.Annotation($"JsonTypeName(\"{model.SerializedName}\")");

                if (hasDerivedModels)
                {
                    javaFile.Line("@JsonSubTypes({");
                    javaFile.Indent(() =>
                    {
                        Func<ClientModel, string> getDerivedTypeAnnotation = (ClientModel derivedType)
                            => $"@JsonSubTypes.Type(name = \"{derivedType.SerializedName}\", value = {derivedType.Name}.class)";

                        foreach (ClientModel derivedModel in model.DerivedModels.SkipLast(1))
                        {
                            javaFile.Line(getDerivedTypeAnnotation(derivedModel) + ',');
                        }
                        javaFile.Line(getDerivedTypeAnnotation(model.DerivedModels.Last()));
                    });
                    javaFile.Line("})");
                }
            }

            if (settings.ShouldGenerateXmlSerialization)
            {
                javaFile.Annotation($"JacksonXmlRootElement(localName = \"{model.XmlName}\")");
            }

            if (model.NeedsFlatten)
            {
                javaFile.Annotation("JsonFlatten");
            }

            List<JavaModifier> classModifiers = new List<JavaModifier>();
            if (!hasDerivedModels && !model.NeedsFlatten)
            {
                classModifiers.Add(JavaModifier.Final);
            }

            string classNameWithBaseType = model.Name;
            if (model.ParentModel != null)
            {
                classNameWithBaseType += $" extends {model.ParentModel.Name}";
            }
            javaFile.Annotation("Fluent");
            javaFile.PublicClass(classModifiers, classNameWithBaseType, (classBlock) =>
            {
                string propertyXmlWrapperClassName(ClientModelProperty property) => property.XmlName + "Wrapper";

                foreach (ClientModelProperty property in model.Properties)
                {
                    string xmlWrapperClassName = propertyXmlWrapperClassName(property);
                    if (settings.ShouldGenerateXmlSerialization && property.IsXmlWrapper)
                    {
                        classBlock.PrivateStaticFinalClass(xmlWrapperClassName, innerClass =>
                        {
                            IType propertyClientType = property.WireType.ClientType;

                            innerClass.Annotation($"JacksonXmlProperty(localName = \"{property.XmlListElementName}\")");
                            innerClass.PrivateFinalMemberVariable(propertyClientType.ToString(), "items");

                            innerClass.Annotation("JsonCreator");
                            innerClass.PrivateConstructor(
                                $"{xmlWrapperClassName}(@JacksonXmlProperty(localName = \"{property.XmlListElementName}\") {propertyClientType} items)",
                                constructor => constructor.Line("this.items = items;"));
                        });
                    }

                    classBlock.BlockComment(settings.MaximumJavadocCommentWidth, (comment) =>
                    {
                        comment.Line(property.Description);
                    });

                    if (!string.IsNullOrEmpty(property.HeaderCollectionPrefix))
                    {
                        classBlock.Annotation("HeaderCollection(\"" + property.HeaderCollectionPrefix + "\")");
                    }
                    else if (settings.ShouldGenerateXmlSerialization && property.IsXmlAttribute)
                    {
                        string localName = settings.ShouldGenerateXmlSerialization ? property.XmlName : property.SerializedName;
                        classBlock.Annotation($"JacksonXmlProperty(localName = \"{localName}\", isAttribute = true)");
                    }
                    else if (settings.ShouldGenerateXmlSerialization && property.WireType is ListType && !property.IsXmlWrapper)
                    {
                        classBlock.Annotation($"JsonProperty(\"{property.XmlListElementName}\")");
                    }
                    else if (!string.IsNullOrEmpty(property.AnnotationArguments))
                    {
                        classBlock.Annotation($"JsonProperty({property.AnnotationArguments})");
                    }

                    if (settings.ShouldGenerateXmlSerialization)
                    {
                        if (property.IsXmlWrapper)
                        {
                            classBlock.PrivateMemberVariable($"{xmlWrapperClassName} {property.Name}");
                        }
                        else if (property.WireType is ListType listType)
                        {
                            classBlock.PrivateMemberVariable($"{property.WireType} {property.Name} = new ArrayList<>()");
                        }
                        else
                        {
                            classBlock.PrivateMemberVariable($"{property.WireType} {property.Name}");
                        }
                    }
                    else
                    {
                        classBlock.PrivateMemberVariable($"{property.WireType} {property.Name}");
                    }
                }

                IEnumerable<ClientModelProperty> constantProperties = model.Properties.Where(property => property.IsConstant);
                if (constantProperties.Any())
                {
                    classBlock.JavadocComment(settings.MaximumJavadocCommentWidth, (comment) =>
                    {
                        comment.Description($"Creates an instance of {model.Name} class.");
                    });
                    classBlock.PublicConstructor($"{model.Name}()", (constructor) =>
                    {
                        foreach (ClientModelProperty constantProperty in constantProperties)
                        {
                            constructor.Line($"{constantProperty.Name} = {constantProperty.DefaultValue};");
                        }
                    });
                }

                foreach (ClientModelProperty property in model.Properties)
                {
                    IType propertyType = property.WireType;
                    IType propertyClientType = propertyType.ClientType;

                    classBlock.JavadocComment(settings.MaximumJavadocCommentWidth, (comment) =>
                    {
                        comment.Description($"Get the {property.Name} property: {property.Description}");
                        comment.Return($"the {property.Name} value");
                    });
                    classBlock.PublicMethod($"{propertyClientType} {property.GetterName}()", (methodBlock) =>
                    {
                        string sourceTypeName = propertyType.ToString();
                        string targetTypeName = propertyClientType.ToString();
                        string expression = $"this.{property.Name}";
                        if (propertyClientType.Equals(ArrayType.ByteArray))
                        {
                            expression = $"ImplUtils.clone({expression})";
                        }
                        if (sourceTypeName == targetTypeName)
                        {
                            if (settings.ShouldGenerateXmlSerialization && property.IsXmlWrapper && property.WireType is ListType listType)
                            {
                                methodBlock.If($"this.{property.Name} == null", ifBlock =>
                                {
                                    ifBlock.Line($"this.{property.Name} = new {propertyXmlWrapperClassName(property)}(new ArrayList<{listType.ElementType}>());");
                                });
                                methodBlock.Return($"this.{property.Name}.items");
                            }
                            else
                            {
                                methodBlock.Return($"{expression}");
                            }
                        }
                        else
                        {
                            methodBlock.If($"{expression} == null", (ifBlock) =>
                            {
                                ifBlock.Return("null");
                            });

                            string propertyConversion = propertyType.ConvertToClientType(expression);

                            methodBlock.Return(propertyConversion);
                        }
                    });

                    if (!property.IsReadOnly)
                    {
                        classBlock.JavadocComment(settings.MaximumJavadocCommentWidth, (comment) =>
                        {
                            comment.Description($"Set the {property.Name} property: {property.Description}");
                            comment.Param(property.Name, $"the {property.Name} value to set");
                            comment.Return($"the {model.Name} object itself.");
                        });
                        classBlock.PublicMethod($"{model.Name} {property.SetterName}({propertyClientType} {property.Name})", (methodBlock) =>
                        {
                            string expression = property.Name;
                            if (propertyClientType.Equals(ArrayType.ByteArray))
                            {
                                expression = $"ImplUtils.clone({expression})";
                            }
                            if (propertyClientType != propertyType)
                            {
                                methodBlock.If($"{property.Name} == null", (ifBlock) =>
                                {
                                    ifBlock.Line($"this.{property.Name} = null;");
                                })
                                .Else((elseBlock) =>
                                {
                                    string sourceTypeName = propertyClientType.ToString();
                                    string targetTypeName = propertyType.ToString();
                                    string propertyConversion = propertyType.ConvertFromClientType(expression);
                                    elseBlock.Line($"this.{property.Name} = {propertyConversion};");
                                });
                            }
                            else
                            {
                                if (settings.ShouldGenerateXmlSerialization && property.IsXmlWrapper)
                                {
                                    methodBlock.Line($"this.{property.Name} = new {propertyXmlWrapperClassName(property)}({expression});");
                                }
                                else
                                {
                                    methodBlock.Line($"this.{property.Name} = {expression};");
                                }
                            }
                            methodBlock.Return("this");
                        });
                    }
                }
            });
        }
    }
}